File: apps/frontend/src/api/budget.ts
/**
 * Budget Management API Client
 * Provides functions to interact with budget tracking endpoints
 */

import { apiClient } from './client';

export interface SessionBudget {
  session_id: string;
  budget_cents: number;
  spent_cents: number;
  remaining_cents: number;
  utilization_percent: number;
  status: 'normal' | 'warning' | 'critical' | 'exhausted';
  transaction_count: number;
  created_at: string;
}

export interface DailyBudget {
  date: string;
  budget_cents: number;
  spent_cents: number;
  remaining_cents: number;
  utilization_percent: number;
  status: 'normal' | 'warning' | 'critical' | 'exhausted';
  transaction_count: number;
  session_count: number;
}

export interface BudgetAnalytics {
  daily: {
    date: string;
    budget_cents: number;
    spent_cents: number;
    remaining_cents: number;
    utilization_percent: number;
    status: string;
    transaction_count: number;
    session_count: number;
  };
  sessions: {
    total_sessions: number;
    active_sessions: number;
    default_budget_cents: number;
    sessions_over_budget: number;
  };
  spending: {
    total_spent_all_time_cents: number;
    total_transactions: number;
    by_model: Record<string, string>;
    avg_cost_per_transaction_cents: number;
  };
}

export interface BudgetAlert {
  timestamp: string;
  alert_type: string;
  message: string;
  utilization_percent: number;
  current_spent_cents: number;
  budget_limit_cents: number;
}

export interface CostForecast {
  session_id: string;
  current_spent_cents: number;
  current_remaining_cents: number;
  total_estimated_cents: number;
  budget_after_tasks_cents: number;
  tasks_breakdown: Array<{
    task_id: string;
    task_name: string;
    estimated_cost_cents: number;
    complexity: number;
  }>;
  budget_sufficient: boolean;
  recommendations: string[];
}

export interface RoutingAnalytics {
  budget: BudgetAnalytics;
  routing: {
    local_complexity_threshold: number;
    paid_complexity_threshold: number;
    default_strategy: string;
    available_local_models: number;
    available_paid_models: number;
  };
}

/**
 * Get session budget details
 */
export const getSessionBudget = async (sessionId: string): Promise<SessionBudget> => {
  const response = await apiClient.get(`/api/v1/budget/session/${sessionId}`);
  return response.data;
};

/**
 * Get daily budget aggregate
 */
export const getDailyBudget = async (): Promise<DailyBudget> => {
  const response = await apiClient.get('/api/v1/budget/daily');
  return response.data;
};

/**
 * Request emergency budget increase
 */
export const requestBudgetIncrease = async (
  sessionId: string,
  amountCents: number,
  justification: string,
  userId: string
): Promise<{
  success: boolean;
  message: string;
  new_budget_cents: number;
  approved_by: string;
}> => {
  const response = await apiClient.post(`/api/v1/budget/session/${sessionId}/increase`, {
    amount_cents: amountCents,
    justification,
    user_id: userId,
  });
  return response.data;
};

/**
 * Get comprehensive budget analytics
 */
export const getBudgetAnalytics = async (): Promise<BudgetAnalytics> => {
  const response = await apiClient.get('/api/v1/budget/analytics');
  return response.data;
};

/**
 * Get cost summary for session or daily
 */
export const getCostSummary = async (
  sessionId: string,
  timeframe: 'session' | 'daily' = 'session'
): Promise<any> => {
  const response = await apiClient.get(`/api/v1/budget/cost-summary/${sessionId}`, {
    params: { timeframe },
  });
  return response.data;
};

/**
 * Get routing analytics
 */
export const getRoutingAnalytics = async (
  sessionId?: string
): Promise<RoutingAnalytics> => {
  const params = sessionId ? { session_id: sessionId } : {};
  const response = await apiClient.get('/api/v1/budget/routing-analytics', { params });
  return response.data;
};

/**
 * Forecast cost for pending tasks
 */
export const forecastCost = async (
  sessionId: string,
  pendingTasks: Array<{
    task_id: string;
    name: string;
    description: string;
    complexity_estimate: number;
    estimated_tokens_input?: number;
    estimated_tokens_output?: number;
  }>
): Promise<CostForecast> => {
  const response = await apiClient.post('/api/v1/budget/forecast', {
    session_id: sessionId,
    pending_tasks: pendingTasks,
  });
  return response.data;
};

/**
 * Get budget alerts for session
 */
export const getBudgetAlerts = async (sessionId: string): Promise<{
  session_id: string;
  alert_count: number;
  alerts: BudgetAlert[];
}> => {
  const response = await apiClient.get(`/api/v1/budget/alerts/${sessionId}`);
  return response.data;
};

/**
 * Manually reset daily budget
 */
export const resetDailyBudget = async (): Promise<{
  success: boolean;
  message: string;
  reset_at: string;
}> => {
  const response = await apiClient.post('/api/v1/budget/reset-daily');
  return response.data;
};

/**
 * Get budget system health status
 */
export const getBudgetHealth = async (): Promise<{
  status: 'healthy' | 'unhealthy';
  components: Record<string, string>;
  daily_budget_status: string;
  daily_utilization_percent: number;
  timestamp: string;
}> => {
  const response = await apiClient.get('/api/v1/budget/health');
  return response.data;
};

File: apps/backend/src/routers/budget.py
"""
Budget Management API Router
Provides endpoints for budget monitoring, analytics, and emergency controls
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from typing import Optional, List, Dict, Any
from pydantic import BaseModel
from datetime import datetime, timezone

from ..core.budget_tracker import get_budget_tracker, BudgetStatus
from ..core.cost_controller import get_cost_controller
from ..core.hybrid_model_router import get_hybrid_router
from ..core.auth import require_roles, ROLE_ADMIN

router = APIRouter(
    prefix="/api/v1/budget",
    tags=["budget"],
    dependencies=[Depends(require_roles(ROLE_ADMIN))],
)


# ============================================================================
# Request/Response Models
# ============================================================================

class BudgetIncreaseRequest(BaseModel):
    """Request to increase budget"""
    amount_cents: int
    justification: str
    user_id: str


class BudgetIncreaseResponse(BaseModel):
    """Response for budget increase"""
    success: bool
    message: str
    new_budget_cents: int
    approved_by: str


class SessionBudgetResponse(BaseModel):
    """Session budget information"""
    session_id: str
    budget_cents: int
    spent_cents: float
    remaining_cents: float
    utilization_percent: float
    status: str
    transaction_count: int
    created_at: str


class DailyBudgetResponse(BaseModel):
    """Daily budget information"""
    date: str
    budget_cents: int
    spent_cents: float
    remaining_cents: float
    utilization_percent: float
    status: str
    transaction_count: int
    session_count: int


# ============================================================================
# Budget Endpoints
# ============================================================================

@router.get("/session/{session_id}", response_model=SessionBudgetResponse)
async def get_session_budget(session_id: str):
    """
    Get current session budget and spending.

    Returns detailed budget information including:
    - Total budget allocation
    - Current spending
    - Remaining budget
    - Utilization percentage
    - Status (normal/warning/critical/exhausted)
    """
    try:
        tracker = get_budget_tracker()
        budget = tracker.get_session_budget(session_id)

        return SessionBudgetResponse(
            session_id=budget.session_id,
            budget_cents=budget.initial_budget_cents,
            spent_cents=budget.spent_cents,
            remaining_cents=budget.remaining_cents,
            utilization_percent=budget.utilization_percent,
            status=budget.status.value,
            transaction_count=budget.transaction_count,
            created_at=budget.created_at.isoformat()
        )

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching session budget: {str(e)}")


@router.get("/daily", response_model=DailyBudgetResponse)
async def get_daily_budget():
    """
    Get current daily budget and spending.

    Returns daily aggregate information across all sessions.
    """
    try:
        tracker = get_budget_tracker()
        budget = tracker.get_daily_budget()

        return DailyBudgetResponse(
            date=budget.date,
            budget_cents=budget.total_budget_cents,
            spent_cents=budget.spent_cents,
            remaining_cents=budget.remaining_cents,
            utilization_percent=budget.utilization_percent,
            status=budget.status.value,
            transaction_count=budget.transaction_count,
            session_count=budget.session_count
        )

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching daily budget: {str(e)}")


@router.post("/session/{session_id}/increase", response_model=BudgetIncreaseResponse)
async def request_budget_increase(session_id: str, request: BudgetIncreaseRequest):
    """
    Request emergency budget increase with approval.

    Small increases (< $5) are auto-approved.
    Larger increases require admin approval.
    """
    try:
        controller = get_cost_controller()

        success, message = await controller.request_emergency_budget(
            session_id=session_id,
            additional_cents=request.amount_cents,
            justification=request.justification,
            user_id=request.user_id
        )

        if not success:
            raise HTTPException(status_code=403, detail=message)

        # Get updated budget
        tracker = get_budget_tracker()
        budget = tracker.get_session_budget(session_id)

        return BudgetIncreaseResponse(
            success=True,
            message=message,
            new_budget_cents=budget.initial_budget_cents,
            approved_by=f"auto_approved_{request.user_id}" if request.amount_cents <= 500 else "pending"
        )

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error processing budget increase: {str(e)}")


@router.get("/analytics")
async def get_budget_analytics():
    """
    Get comprehensive budget analytics.

    Returns:
    - Daily and session summaries
    - Spending breakdown by model
    - Transaction statistics
    - Cost trends
    """
    try:
        tracker = get_budget_tracker()
        analytics = await tracker.get_budget_analytics()

        return analytics

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching analytics: {str(e)}")


@router.get("/cost-summary/{session_id}")
async def get_cost_summary(
    session_id: str,
    timeframe: str = Query("session", regex="^(session|daily)$")
):
    """
    Get cost summary for a session or daily.

    Query Parameters:
    - timeframe: "session" or "daily"
    """
    try:
        controller = get_cost_controller()
        summary = await controller.get_cost_summary(
            session_id=session_id if timeframe == "session" else None,
            timeframe=timeframe
        )

        return summary

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching cost summary: {str(e)}")


@router.get("/routing-analytics")
async def get_routing_analytics(session_id: Optional[str] = None):
    """
    Get routing analytics and model selection statistics.

    Shows how tasks are being routed between local and paid models.
    """
    try:
        router_instance = get_hybrid_router()
        analytics = await router_instance.get_routing_analytics(session_id)

        return analytics

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching routing analytics: {str(e)}")


@router.post("/reset-daily")
async def reset_daily_budget():
    """
    Manually trigger daily budget reset.

    Normally runs automatically at midnight UTC.
    This endpoint allows manual reset for testing or emergency situations.
    """
    try:
        tracker = get_budget_tracker()
        success, message = await tracker.reset_daily_budget()

        if not success:
            raise HTTPException(status_code=500, detail=message)

        return {
            "success": True,
            "message": message,
            "reset_at": datetime.now(timezone.utc).isoformat()
        }

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error resetting daily budget: {str(e)}")


@router.get("/health")
async def budget_health_check():
    """
    Health check for budget tracking system.

    Returns status of budget tracker, cost controller, and hybrid router.
    """
    try:
        tracker = get_budget_tracker()
        controller = get_cost_controller()
        router_instance = get_hybrid_router()

        daily = tracker.get_daily_budget()

        return {
            "status": "healthy",
            "components": {
                "budget_tracker": "operational",
                "cost_controller": "operational",
                "hybrid_router": "operational"
            },
            "daily_budget_status": daily.status.value,
            "daily_utilization_percent": daily.utilization_percent,
            "timestamp": datetime.now(timezone.utc).isoformat()
        }

    except Exception as e:
        return {
            "status": "unhealthy",
            "error": str(e),
            "timestamp": datetime.now(timezone.utc).isoformat()
        }


# ============================================================================
# Cost Forecasting Endpoints
# ============================================================================

class TaskEstimate(BaseModel):
    """Task for cost estimation"""
    task_id: str
    name: str
    description: str
    complexity_estimate: int
    estimated_tokens_input: Optional[int] = None
    estimated_tokens_output: Optional[int] = None


class ForecastRequest(BaseModel):
    """Request for cost forecast"""
    session_id: str
    pending_tasks: List[TaskEstimate]


@router.post("/forecast")
async def forecast_session_cost(request: ForecastRequest):
    """
    Forecast total cost for pending tasks.

    Provides:
    - Estimated cost per task
    - Total estimated cost
    - Budget sufficiency check
    - Recommendations for optimization
    """
    try:
        from ..core.model_bidding import TaskDefinition

        controller = get_cost_controller()

        # Convert TaskEstimate to TaskDefinition
        tasks = []
        for task_est in request.pending_tasks:
            task_def = TaskDefinition(
                task_id=task_est.task_id,
                name=task_est.name,
                description=task_est.description,
                complexity_estimate=task_est.complexity_estimate,
                required_capabilities=["reasoning"],
                security_sensitive=False,
                estimated_tokens_input=task_est.estimated_tokens_input or 2000,
                estimated_tokens_output=task_est.estimated_tokens_output or 1000
            )
            tasks.append(task_def)

        # Get forecast
        forecast = await controller.forecast_session_cost(
            session_id=request.session_id,
            pending_tasks=tasks
        )

        return forecast

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error generating forecast: {str(e)}")


# ============================================================================
# Budget Alerts Endpoint
# ============================================================================

@router.get("/alerts/{session_id}")
async def get_budget_alerts(session_id: str):
    """
    Get budget alerts for a session.

    Returns all warning and critical alerts issued.
    """
    try:
        tracker = get_budget_tracker()
        budget = tracker.get_session_budget(session_id)

        alerts = [
            {
                "timestamp": alert.timestamp.isoformat(),
                "alert_type": alert.alert_type,
                "message": alert.message,
                "utilization_percent": alert.utilization_percent,
                "current_spent_cents": alert.current_spent_cents,
                "budget_limit_cents": alert.budget_limit_cents
            }
            for alert in budget.alerts_sent
        ]

        return {
            "session_id": session_id,
            "alert_count": len(alerts),
            "alerts": alerts
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching alerts: {str(e)}")
